libs/core/langchain_core/runnables/context.py

```diff
@@ -0,0 +1,259 @@
+from langchain_core.runnables.utils import ConfigurableFieldSpec, Input, Output
+
+T = TypeVar("T")
+Values = Dict[Union[asyncio.Event, threading.Event], Any]
+CONTEXT_CONFIG_PREFIX = "__context__/"
+CONTEXT_CONFIG_SUFFIX_GET = "/get"
+CONTEXT_CONFIG_SUFFIX_SET = "/set"
+
+
+async def _asetter(done: asyncio.Event, values: Values, value: T) -> T:
+    values[done] = value
+    done.set()
+    return value
+
+
+async def _agetter(done: asyncio.Event, values: Values) -> Any:
+    await done.wait()
+    return values[done]
+
+
+def _setter(done: threading.Event, values: Values, value: T) -> T:
+    values[done] = value
+    done.set()
+    return value
+
+
+def _getter(done: threading.Event, values: Values) -> Any:
+    done.wait()
+    return values[done]
+
+
+def _config_with_context(
+    config: RunnableConfig,
+    steps: List[Runnable],
+    setter: Callable,
+    getter: Callable,
+    event_cls: Union[Type[threading.Event], Type[asyncio.Event]],
+) -> RunnableConfig:
+    if any(k.startswith(CONTEXT_CONFIG_PREFIX) for k in config.get("configurable", {})):
+        return config
+
+    context_specs = [
+        (spec, i)
+        for i, step in enumerate(steps)
+        for spec in step.config_specs
+        if spec.id.startswith(CONTEXT_CONFIG_PREFIX)
+    ]
+    grouped_by_key = groupby(
+        sorted(context_specs, key=lambda s: s[0].id),
+        key=lambda s: s[0].id.split("/")[1],
+    )
+
+    values: Values = {}
+    events: DefaultDict[str, Union[asyncio.Event, threading.Event]] = defaultdict(
+        event_cls
+    )
+    context_funcs: Dict[str, Callable[[], Any]] = {}
+    for key, group in grouped_by_key:
+        lgroup = list(group)
+        getters = [s for s in lgroup if s[0].id.endswith(CONTEXT_CONFIG_SUFFIX_GET)]
+        setters = [s for s in lgroup if s[0].id.endswith(CONTEXT_CONFIG_SUFFIX_SET)]
+
+        if len(getters) < 1:
+            raise ValueError(f"Expected at least one getter for context key {key}")
+        if len(setters) != 1:
+            raise ValueError(f"Expected exactly one setter for context key {key}")
+        setter_idx = setters[0][1]
+        if any(getter_idx < setter_idx for _, getter_idx in getters):
+            raise ValueError(
+                f"Context setter for key {key} must be defined after all getters."
+            )
+
+        context_funcs[getters[0][0].id] = partial(getter, events[key], values)
+        context_funcs[setters[0][0].id] = partial(setter, events[key], values)
+
+    return patch_config(config, configurable=context_funcs)
+
+
+def aconfig_with_context(
+    config: RunnableConfig,
+    steps: List[Runnable],
+) -> RunnableConfig:
+    return _config_with_context(config, steps, _asetter, _agetter, asyncio.Event)
+
+
+def config_with_context(
+    config: RunnableConfig,
+    steps: List[Runnable],
+) -> RunnableConfig:
+    return _config_with_context(config, steps, _setter, _getter, threading.Event)
+
+
+class ContextGet(RunnableSerializable):
```

Comment:
Would it be possible to have context get support remapping of keys? It's not uncommon to have to rename keys when writing chains

```Context.getter(foo='input')``` could this do 

```
{
  'foo': configurable[ id for input ] () ?
}
```

----

IMO would be nice to have a similar feature in `Context.setter` if the semantics about what the return value is can be made clear